### PR TITLE
Wait for effective publishing permission on new feeds

### LINKED ISSUE #17617: Wait for effective Contributor permission before publishing to newly created feeds
## Description
Stable build promotion creates dedicated Azure Artifacts feeds and immediately publishes packages to them. `CreateAzureDevOpsFeed.WaitForFeedReadyAsync` currently considers the feed ready when the publishing credential can GET the NuGet service index. That proves readability, but not that the newly assigned Contributor role has propagated to the package upload authorization path.

This caused two production promotion failures on September 23, 2026:

- macios DevDiv Build Promotion build 15431760
- Aspire dnceng Build Promotion builds 3085623 and 3085650

In the Aspire failure, the feed was declared ready 472 ms after creation. Package uploads began about 16.6 seconds later and failed with `403 FeedNeedsPermissionsException`: `maestro-build-promotion-mi` lacked `AddPackage`. The retry created a suffixed feed and reproduced the failure. Inspection showed the new feed still reported the managed identity role as `none`; the earlier feed eventually reported `contributor`.

## Root cause

Azure Artifacts feed creation and effective permission propagation are asynchronous. A successful service-index GET is not a publishing-readiness signal, especially for public feeds that are readable without Contributor access.

## Proposed fix

For feeds created with explicit publisher permissions, poll the feed permissions endpoint until every configured publishing identity has its expected Contributor role. Retain the existing readability fallback for feeds that do not define explicit permissions. Add regression tests proving that a readable feed is not considered publish-ready while the publisher role is still `none`.

### PR DESCRIPTION
**Pull request by missymessa**
## Summary

- poll the Azure Artifacts feed permissions endpoint until configured publisher roles are effective
- do not treat public or inherited read access to the NuGet service index as publish readiness
- preserve the existing service-index readiness fallback for feeds without explicit publisher permissions
- add regression coverage for missing, `none`, read-only, and effective Contributor states

Fixes #17617

## Validation

- `CreateAzureDevOpsFeedTests`: 4 passed
- affected project builds successfully

The full test project runs an unrelated symbol publishing test that requires Windows x64; this checkout is running Windows ARM64.

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs [hunk 1]
```diff
@@ -138,6 +144,92 @@ public async Task<bool> WaitForFeedReadyAsync(
...
+                    string responseBody = await response.Content.ReadAsStringAsync();
+                    AzureDevOpsFeedPermissionsResponse permissionsResponse =
+                        JsonConvert.DeserializeObject<AzureDevOpsFeedPermissionsResponse>(responseBody);
+                    lastPermissions = permissionsResponse?.Value ?? Array.Empty<AzureDevOpsFeedPermissionResponse>();
+
+                    bool allPermissionsEffective = requiredPermissions.All(required =>
+                        lastPermissions.Any(actual =>
+                            actual.IsInheritedRole &&
+                            string.Equals(actual.IdentityDescriptor, required.IdentityDescriptor, StringComparison.OrdinalIgnoreCase) &&
+                            string.Equals(actual.Role, required.Role, StringComparison.OrdinalIgnoreCase)));
```
[the remark this answers is not in the document]
Comment (missymessa) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs:179]: Thanks for flagging this. Live validation showed that Azure Artifacts returns two entries for the same descriptor during propagation: the requested direct assignment (`contributor`, `isInheritedRole: false`) and a computed entry (`none`/`reader`, `isInheritedRole: true`). Accepting the direct entry reproduced the premature-ready failure: the task returned ready while the effective entry was still `none`, and `AddPackage` returned 403.

The regression test deliberately models that sequence and now has a more explicit name (`WaitForFeedPermissionsReadyIgnoresDirectContributorUntilComputedRoleIsEffective`). Commit 42c99ed08 also documents the observed response behavior at the check. I am retaining the `isInheritedRole` filter because removing it would restore the live failure this PR is intended to prevent.

A separate canary also showed that package authorization can briefly lag even after the computed role reaches `contributor`; I am treating actual package-push retry as an additional requirement before this PR is merge-ready.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs [hunk 2]
```diff
@@ -138,6 +144,95 @@ public async Task<bool> WaitForFeedReadyAsync(
...
+                            // The API can return the requested direct assignment (false) while its
+                            // computed role is still "none" or "reader". Only the computed entry
+                            // reflected by isInheritedRole predicts authorization propagation.
+                            actual.IsInheritedRole &&
+                            string.Equals(actual.IdentityDescriptor, required.IdentityDescriptor, StringComparison.OrdinalIgnoreCase) &&
+                            string.Equals(actual.Role, required.Role, StringComparison.OrdinalIgnoreCase)));
+
+                    if (allPermissionsEffective)
+                    {
+                        return RetryResult.Success;
```
[the remark this answers is not in the document]
Comment (missymessa) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs:183]: Addressed in 999c916fa. Package uploads now classify Azure Artifacts HTTP 401/403 responses separately and keep those authorization failures retryable for up to 12 attempts, with each delay capped at two minutes (about an 11–13 minute bounded window with the default backoff). Generic upload failures and package-content checks retain the existing five-attempt limit.

The extended path applies to all Azure DevOps package feeds rather than `Isolated` feeds because `PublishSignedAssets` can push to a feed created moments earlier without setting `TargetFeedConfig.Isolated`. Tests cover 401/403 classification, success on the final authorization attempt, exhaustion after 12 attempts, and preservation of the existing retry behavior. The focused suite passes 19/19.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs [hunk 3]
```diff
@@ -249,13 +348,41 @@ private async Task<bool> ExecuteAsync()
             using (HttpClient readinessClient = new HttpClient(new HttpClientHandler { CheckCertificateRevocationList = true }))
             {
                 readinessClient.DefaultRequestHeaders.Authorization = GeneralUtils.CreateAzdoAuthHeader(AzureDevOpsPersonalAccessToken);
-                if (!await WaitForFeedReadyAsync(TargetFeedURL, readinessClient, FeedReadinessRetryHandler))
+                bool feedIsReady;
+                if (publishingPermissions.Count > 0)
```
Comment (mmitche) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateAzureDevOpsFeed.cs]: publishingPermissions should always be non-empty, unless you can't see the feed, so I think this won't do what we want.

We should first wait for feed existence, using the timeout. That will imp;licitly will check perms for some feeds. Then, check for a specific add package permission.

**Review by mmitche — CHANGES_REQUESTED** [review]
Looks a bit off. Please see comment.
